pallets/settlement-risc0/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+    use frame_system::pallet_prelude::OriginFor;
+    use hp_poe::OnProofVerified;
+    use sp_core::H256;
+    use sp_io::hashing::keccak_256;
+    use sp_std::{boxed::Box, vec::Vec};
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    // Errors inform users that something went wrong.
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Provided proof is not valid.
+        InvalidProof,
+        /// Provided proof has excessive size.
+        InvalidProofSize,
+        /// Provided public inputs are not valid.
+        InvalidPublicInputs,
+        /// Provided public inputs have excessive size.
+        InvalidPublicInputsSize,
+        /// Verify proof failed.
+        VerifyError,
+    }
+
+    impl<T> From<native::VerifyError> for Error<T> {
+        fn from(e: native::VerifyError) -> Self {
+            match e {
+                native::VerifyError::InvalidInput => Error::<T>::InvalidPublicInputs,
+                native::VerifyError::InvalidProofData => Error::<T>::InvalidProof,
+                native::VerifyError::VerifyError => Error::<T>::VerifyError,
+            }
+        }
+    }
+
+    /// Configure the pallet by specifying the parameters and types on which it depends.
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Proof verified call back
+        type OnProofVerified: OnProofVerified;
+        /// The weight definition for this pallet
+        type WeightInfo: WeightInfo;
+    }
+
+    pub fn verify_proof<T: Config>(
+        vk: [u8; 32],
+        proof: Vec<u8>,
+        pubs: Vec<u8>,
+    ) -> Result<(), Error<T>> {
+        log::trace!("Checking size");
+        ensure!(
+            (proof).len() <= MAX_PROOF_SIZE,
+            Error::<T>::InvalidProofSize
+        );
+        ensure!(
+            (pubs).len() <= MAX_PUBS_SIZE,
+            Error::<T>::InvalidPublicInputsSize
+        );
+
+        log::trace!("Verifying (native)");
+        native::risc_0_verify::verify(vk, &proof, &pubs)
+            .map(|_x| T::OnProofVerified::on_proof_verified(compute_risc0_hash(vk, pubs)))
+            .map_err(Into::into)
+    }
+
+    // Dispatchable functions allows users to interact with the pallet and invoke state changes.
+    // These functions materialize as "extrinsics", which are often compared to transactions.
+    // Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+    #[pallet::call(weight(<T as Config>::WeightInfo))]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        pub fn submit_proof(
+            _origin: OriginFor<T>,
+            vk_u8: [u8; 32],
+            proof: Box<Vec<u8>>,
+            pubs: Box<Vec<u8>>,
+        ) -> DispatchResultWithPostInfo {
+            log::trace!("Submitting proof");
+            verify_proof::<T>(vk_u8, *proof, *pubs)
+                .map(Into::into)
+                .map_err(Into::into)
+        }
+    }
+
```

Comment:
You can remove all `Box` here: take care of remove also all into inthe related calls